apps/src/p5lab/spritelab/poembot/commands/foregroundEffects.js

```diff
@@ -5,66 +5,91 @@ import drawStar from '@code-dot-org/dance-party/src/shapes/star';
 import drawStarstruck from '@code-dot-org/dance-party/src/shapes/starstruck';
 import drawTickled from '@code-dot-org/dance-party/src/shapes/tickled';
 import drawWink from '@code-dot-org/dance-party/src/shapes/wink';
+import {createUuid} from '@cdo/apps/utils';
 import * as utils from './utils';
 
 export const commands = {
   setForegroundEffect(effectName) {
+    const id = createUuid();
     switch (effectName) {
       case 'rain': {
-        const drops = [];
+        let drops = [];
         for (let i = 0; i < 20; i++) {
           drops.push({
-            x: utils.randomInt(0, 380),
-            y: utils.randomInt(0, 380),
+            x: utils.randomInt(-400, 380),
+            y: utils.randomInt(-50, -20),
             length: utils.randomInt(10, 20)
           });
         }
-        this.foregroundEffect = () => {
-          this.p5.push();
-          this.p5.stroke(this.p5.rgb(92, 101, utils.randomInt(140, 220), 0.5));
-          this.p5.strokeWeight(3);
-          for (let i = 0; i < drops.length; i++) {
+        this.foregroundEffects.push({
+          id,
+          func: () => {
             this.p5.push();
-            this.p5.translate(drops[i].x - 20, drops[i].y - 20);
-            this.p5.line(0, 0, drops[i].length, drops[i].length * 2);
-            drops[i].y = (drops[i].y + drops[i].length) % 420;
-            drops[i].x = (drops[i].x + drops[i].length / 2) % 420;
+            this.p5.stroke(
+              this.p5.rgb(92, 101, utils.randomInt(140, 220), 0.5)
+            );
+            this.p5.strokeWeight(3);
+            for (let i = 0; i < drops.length; i++) {
+              this.p5.push();
+              this.p5.translate(drops[i].x - 20, drops[i].y - 20);
+              this.p5.line(0, 0, drops[i].length, drops[i].length * 2);
+              drops[i].y = drops[i].y + drops[i].length;
+              drops[i].x = drops[i].x + drops[i].length / 2;
+              this.p5.pop();
+            }
+            drops = drops.filter(drop => drop.y < 420 && drop.x < 420);
+            if (drops.length === 0) {
+              // Remove this foreground effect once no more drops are visible.
+              this.foregroundEffects = this.foregroundEffects.filter(
+                effect => effect.id !== id
+              );
+            }
             this.p5.pop();
           }
-          this.p5.pop();
-        };
+        });
         break;
       }
       case 'bubbles': {
         let bubbles = [];
-        this.foregroundEffect = () => {
+        for (let i = 0; i < 25; i++) {
```

Comment:
nit for this effect and others that now generate a set number of items in the effect: we should pull this value into a constant so we can easily identify/change how many bubbles are generated as part of the effect